README

ADMS - An automatic device model synthesizer.

Forked from the source code available at: http://sourceforge.net/projects/mot-adms/files/adms-source/2.3/
The original SVN repo is now password-protected, so here we are.

## Autotools install:

You should be able to build with:
    ./configure --prefix=/install/location/
    make
    make install


There was an INSTALL file, but I removed it because zsh kept suggesting it as a correction when running
`make install`. If you find that you really need it, you can replace it by running: `automake --add-missing`


## CMake install:

Requires bison (2.6+), flex.
Tested on Mac OS X and Ubuntu 12.04.
Changes for other systems should probably be done in admsXml/CMakeLists.txt and admsXml/config.h.

To install:

  bash create_files.sh
  mkdir cmake
  cd cmake
  cmake .. -DCMAKE_BUILD_TYPE=RELEASE -DCMAKE_INSTALL_PREFIX=/install/location/
  make install


---

Original author's build instructions: (for posterity)


0- Prerequisites
    0- gnu make is required (do not use other native make)
 
1- Description
  ADMS is a code generator that converts electrical compact device models
  specified in high-level description language into ready-to-compile c code
  for the API of spice simulators.

2- Installation - Unix or Linux
    0- run: ./configure - see file INSTALLATION for more options
    1- run: gmake
    2- run: gmake install
  If the installation fails try:
    0- run: autoheader
    1- run: aclocal
    2- run: automake
    3- run: autoconf
    4- run: ./configure - see file INSTALLATION for more options
    5- run: gmake
    6- run: gmake install
    In two shots:
      autoheader && aclocal && automake && autoconf && ./configure
      gmake && gmake install 

3- Installation - Windows with cygwin environment
    0- run: ./configure - see file INSTALLATION for more options
    1- run: gmake
    2- run: gmake install
  If the installation fails try:
    0- run: autoheader
    1- run: aclocal
    2- run: automake
    3- run: autoconf
    4- run: ./configure - see file INSTALLATION for more options
    5- run: gmake
    6- run: gmake install
    In two shots:
      autoheader && aclocal && automake && autoconf && ./configure
      gmake && gmake install 

4- Installation windows (no cygwin environment)
  0- run:
       cl -DYY_NO_UNISTD_H -DPACKAGE_VERSION=\"msvc-229\" -DHAVE_CONFIG_H -DinsideElement -DinsidePreprocessor -DinsideVeriloga -I. -IadmsXml \
          admsXml/adms.c admsXml/preprocessorLex.c admsXml/preprocessorMain.c admsXml/preprocessorYacc.c admsXml/verilogaLex.c admsXml/verilogaYacc.c \
          admsXml/admsXml.c admsXml/admstpathYacc.c -o admsXml.exe
     If the installation fails send a message to r29173@users.sourceforge.net
     (note that there is NO strong support for windows compilers.)

5- Directory Structure
   adms
    - metaadms.dtd
      DTD of file adms.xml 
    - adms.xml
      This file defines the data structure used by adms to save parsed hdl code.
    - auxconf
      set-up files created by automake -a -c
    - scripts:
      perl scripts
    - admsXml
      source code

5- For developers (using the cygwin environment)
  Here is how I create the compilation environment after checking out adms from
  the CVS repository:
  0- run: autoheader (create autom4te.cache/config-h.in from configure.in)
  1- run: mkdir auxconf (create directory auxconf)
  2- run: aclocal (create aclocal.m4 from configure.in)
  3- run: libtoolize --force --ltdl -c (create libltdl and config.guess, config.sub, ltmain.sh in auxconf)
  4- run: automake -a -c (create missing, mkinstalldirs, install-sh in auxconf and all Makefile.in)
  5- run: autoconf (create configure)
  6- run: ./configure --enable-maintainer-mode
  In three shots:
    rm -rf auxconf && autoheader && mkdir auxconf && aclocal && libtoolize --force --ltdl -c 
    touch ChangeLog && automake -a -c && autoconf
    ./configure --enable-maintainer-mode



NEWS
#####################################################################
#                      ChangeLog                                    #
#####################################################################

10-10-2000: Increment version 1.0.2
            1.0.1 has been used to compile r3-1.0.0
01-06-2001: Increment version 1.0.3
            Made a lot of simplifications 
            Remove class admsSemantic
            Add module admsNS
            Test-suite golden with previous SUCCEED
18-07-2001: Increment version 1.0.4
            Made a lot of simplifications
            Test-suite golden with previous SUCCEED
25-07-2001: Increment version 1.0.5
	    mica/loadAC calls loadTRAN instead loadDC
            Test-suite golden with previous FAILED (mica)
30-07-2001: Increment version 1.0.6
	    useless classes removed
            Test-suite golden with previous SUCCEED
10-08-2001: Increment version 1.0.7
            Made a lot of simplifications
            Test-suite golden with previous SUCCEED
13-08-2001: Increment version 1.0.8
            Alias are not printed out by simulators
            Test-suite golden with previous FAILED (mica/spectre)
07-11-2002: Increment version: now 1.0.24
            mica models created with this version (SP, MOSCAP3, R3, VBICSELFT)
07-11-2002: Increment version: now 1.0.25
            mica models: 1.0.24 and 1.0.25 generate same c code for:
            HBT-1 MOSCAP3-9 R3-13 SMOS-16 SP-16 VBICSELFT-23 MOS1_lilly_qi-0
            handles arrays of variables: module PDSOI_lilly_qi-0
20-11-2002: Increment version: now 1.1.0
            start implementing admsXml
            mica models: 1.1.0 and 1.0.25 generate same c code
01-12-2002: Increment version: now 1.1.1
            improve  admsXml
            mica models: 1.1.1 and 1.0.25 generate same c code
            except for adms*/data/*interface_h.in (fix to _d_pow)
08-01-2003: Increment version: now 1.1.2
            mica models: 1.1.2 and 1.1.1 generate same c code
            except MICAinterface.h (fixes to _limexp)
            except *defs.h and *noise.c (started implementation of noise analysis)
19-03-2003: Increment version: now 1.1.3
19-03-2003: Increment version: now 1.1.4
            mica models: 1.1.4 and 1.1.2 generate same c code except
              file <MODULE>reverse.c (one comment added)
              module VBICSELFT: 
		VBICSELFTdefs.h
		VBICSELFTevaluate.include
		VBICSELFTload.include
		VBICSELFTreserve.c
              module MOS1: it is a `A' device 
19-03-2003: Increment version: now 1.1.5
            mica models: 1.1.5 and 1.1.4 generate same c code except
            xml scripts for spectre 90% tested
            TODO: 1- "flag parameters" should be processed like
                     real/int/string operators
                  2- remove "ask=" field
                  3- remove GLOBAL flag
                  4- add NOISE
                  5- add xpath=/adms to xml scripts

08-04-2003: Increment version: now 1.1.6
            mica models: 1.1.6 and 1.1.5 generate same c code except
            WarningL R3-16 does not compile with this version.
            mica5HBT-1/HBTtopo.c mica6HBT-1/HBTtopo.c
            mica5MOSCAP3-14/MOSCAP3topo.c mica6MOSCAP3-14/MOSCAP3topo.c
            mica5VBICSELFT-24/VBICSELFTtopo.c mica6VBICSELFT-24/VBICSELFTtopo.c
            Same contents, however order of lines differ
            Keep changes and move to next version
11-04-2003: Increment version: now 1.1.7
            mica models: 1.1.7 and 1.1.6 generate same c code for mica
            Did lot of clean-up in the code
15-04-2003: Increment version: now 1.1.8
            mica models: 1.1.8 and 1.1.7 generate same c code for mica
            Did lot of clean-up
16-04-2003: Increment version: now 1.1.9
            mica models: 1.1.9 and 1.1.8 generate same c code for mica.
            This version does not compile on Windows
17-04-2003: Increment version: now 1.1.10
            mica models: 1.1.10 and 1.1.8 generate same c code
            for mica/spectre/ads
            Fixes to make adms compiled on Windows
            No check done on HPUX/SUN
18-04-2003: Increment version: now 1.1.11
            fix bug in admsCheck that prevented use of gprof
            mica models: 1.1.11 and 1.1.8 generate same c code
            for mica/spectre/ads
18-05-2003: Increment version: now 1.1.12
            fix bug in admsCheck that prevented use of gprof
            spectre: handles parameter range checking
            ads: improve syntax of code created
            mica: same c code as 1.1.8
03-05-2005: removed admsNS directory
            release adms-2.0.1
===== gap ====
21-11-2016: 2.3.6
18-08-2020: 2.3.7. minor release.
            reproducible build, typo fixes,
            include path, instructions
*****-2022: 2.3.8.
            add more numerical constants
            provide constants script for more flexibility
            fix -I, take precedence over ADMS_INCLUDEDIR
            allow semicolon after identifier in disciple/nature declaration
            add basic support for escaped identifiers
            resolve include directives relative to each file


admsXml/admsCheck.c
/*
This file is part of adms - http://sourceforge.net/projects/mot-adms.

adms is a code generator for the Verilog-AMS language.

Copyright (C) 2002-2012 Laurent Lemaitre <r29173@users.sourceforge.net>
              2013 Richard Crozier <richard.crozier@example.com>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

/*
 * RCS Info
 * $Id: admsCheck.c 1110 2008-10-14 13:40:29Z r29173 $
 *
 * Log
 * $Log$
 * Revision 1.17  2006/09/13 16:13:20  r29173
 * started migration to www.antlr.org tools
 *
 * Revision 1.16  2006/08/04 17:15:32  r29173
 * rearranged code partionning into files
 *
 * Revision 1.15  2006/06/16 12:24:25  r29173
 * ton of changes: memory clean-up
 *
 * Revision 1.14  2006/06/09 16:49:34  r29173
 * improved malloc management
 *
 * Revision 1.13  2006/01/26 13:39:25  r29173
 * fixed core dump when running admsCheck
 *
 * Revision 1.12  2005/06/21 08:02:28  r29173
 * renamed adms_set_admsmain to adms_valueto_admsmain
 *
 * Revision 1.11  2005/05/09 14:38:29  r29173
 * cleaned-up source directory structure
 *
 * Revision 1.10  2005/05/03 09:35:14  r29173
 * cleaned-up header file dependencies
 *
 * Revision 1.9  2005/05/03 07:58:19  r29173
 * moved admsNS into admsFile
 *
 * Revision 1.8  2005/04/01 07:25:44  r29173
 * renamed prefix .xsl to .xml
 *
 * Revision 1.7  2005/01/11 09:58:01  r29173
 * tons of changes - simplify accessor names (valueof, valueto)
 *
 * Revision 1.6  2004/11/19 12:29:56  r29173
 * renamed _C into CAT
 * saved argc, argv into admsmain element
 * added adms:setenv transform
 *
 * Revision 1.5  2004/11/11 13:45:03  r29173
 * replaced g_message by adms_message_info
 *
 * Revision 1.4  2004/11/01 09:37:36  r29173
 * changed messaging macros (avoid useless computation of arguments)
 * improved the coding of transforms in xmlProgram.c
 *
 * Revision 1.3  2004/08/03 12:33:54  r29173
 * import adms-1.21.0 from local CVS
 *
 * Revision 1.2  2004/08/03 10:05:46  r29173
 * removed superelement from admsgui
 * fixed some warnings issued during compilation
 *
 * Revision 1.1.1.1  2004/05/21 12:20:01  r29173
 * recreated cvs data structure (crashed after revision 1.13.0!)
 *
 * Revision 1.10  2004/03/11 16:25:14  r29173
 * started implementing code builder in yacc parser
 *
 * Revision 1.9  2004/03/08 13:58:05  r29173
 * all code lower-cased
 *
 * Revision 1.8  2004/03/08 08:24:19  r29173
 * started implementing code builder in yacc parser
 *
 * Revision 1.7  2004/01/13 11:25:51  r29173
 * added adms_message_usage
 *
 * Revision 1.6  2004/01/09 16:23:45  r29173
 * cleaned-up use of #include
 *
 * Revision 1.5  2004/01/08 10:28:49  r29173
 * moved code to always.[ch]
 *
 * Revision 1.4  2004/01/05 11:07:51  r29173
 * change the way admsCheck.xml is downloaded
 *
 * Revision 1.3  2003/10/22 15:11:36  r29173
 * changed the way script "text2Ccode.pl" handles file name.
 *
 * Revision 1.2  2003/05/21 14:18:00  r29173
 * add rcs info
 *
 */


/*
  admsCheck

  Small program to perform a test run of adms using dummy.va and dummy.xml

  Used by 'make check' in the adms build system

*/

#include "adms.h"
/*constants*/
#include "dummy.xml.c"
#include "dummy.va.c"

char*mydata[]={
"dummy.va","dummy_va",
"dummy.xml","dummy_xml",
NULL
};

int main(const int argc, const char **argv)
{
  char**mydataptr;
  rootnew(adms_admsmain_new("admsmain"));
  adms_admsmain_valueto_filename(root(),adms_kclone("stdin"));
  adms_message_info(("Create files:\n"));
  for( mydataptr=mydata; *mydataptr; mydataptr++,mydataptr++)
  {
    char*savefilename=*mydataptr;
    char*symbolname=*(mydataptr+1);
    char* myfilename=strdup(savefilename);
    FILE* ofh=fopen(myfilename,"w");
    adms_message_info((" ... %s\n",myfilename));
    free(myfilename);
    if(!strcmp(symbolname,"dummy_va"))
      fprintf(ofh,"%s",dummy_va);
    else if(!strcmp(symbolname,"dummy_xml"))
      fprintf(ofh,"%s",dummy_xml);
    fclose(ofh);
  }
  adms_message_info(("How to use the example\n"));
  adms_message_info(("  admsXml dummy.va -e dummy.xml\n"));
  adms_message_info(("  This command parses file dummy.va\n"));
  adms_message_info(("  and interprets control file dummy.xml\n"));
  adms_message_info(("  File dummy.net is created.\n"));
  adms_message_info(("  (file .dummy.va.adms is created in background.)\n"));
  return 0;
}
